File: ABC340/d.py
n = int(input())
inf = 1000000000000000
s_time = [inf]*(n+1)
#print(s_time)
s_time[0] = 0
s_time[1] = 0
for i in range(1, n):
    a, b, x = map(int, input().split())
    s_time[i+1] = min(s_time[i+1], a + s_time[i])
    s_time[x] = min(s_time[x], b + s_time[i])

print(s_time[n])
print(s_time)

File: ABC340/d.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<algorithm>
using namespace std;

int main(){
    int n;
    cin >> n;
    int a, b, x;
    long long INF = 1000000000000000;
    vector<vector<pair<int, int>>> g(n);
    vector<long long> time(n, INF);
    vector<bool> is_visited(n, false);
    priority_queue<pair<long long, int>, vector<pair<long long, int>>, greater<pair<long long, int>>> next_v_que;

    for(int i=0; i<n-1; i++){
        cin >> a >> b >> x;
        x--;
        g[i].push_back({i+1, a});
        g[i].push_back({x, b});
    }
    
    time[0] = 0;
    next_v_que.push({time[0], 0});

    while(!next_v_que.empty()){
        auto [now_t, now_v] = next_v_que.top();
        next_v_que.pop();
        if(is_visited[now_v]) continue;
        is_visited[now_v] = true;

        for(auto [child_v, child_w]:g[now_v]){            
            if(is_visited[child_v]) continue;
            time[child_v] = min(time[child_v], time[now_v]+child_w);
            next_v_que.push({time[child_v], child_v});
        }
    }

    cout << time[n-1] << endl;

    return 0;
}